typegate/src/planner/args.ts

```diff
@@ -816,14 +828,33 @@ class TypeMismatchError extends Error {
   }
 }
 
-const ajv = new Ajv();
+const visitedObject = new Set<string>();
+const ajv = new Ajv({
+  allowUnionTypes: true,
+});
 addFormats(ajv);
 addJsonFormat(ajv);
 
 function buildJsonSchema(typeNode: TypeNode, tg: TypeGraph) {
   return new JsonSchemaBuilder(tg).build(typeNode);
 }
 
+function pathOf(typeNode: TypeNode): string {
+  return `#/schemas/temp/${typeNode.title}`;
+}
+
+function refOf(typeNode: TypeNode): JSONSchema {
+  return { $ref: pathOf(typeNode) };
+}
+
+function addSchemaToAjv(schema: JSONSchema, path: string) {
+  if (ajv.getSchema(path)) return;
+  try {
+    ajv.addSchema(schema, path);
+    ajv.compile(schema); // cache if possible
+  } catch (_err) { /* ref cannot be resolved yet */ }
```

Comment:
`Type.OBJECT` is the only type cached. In fact, it still works if I remove `ajv.compile(.)` in this function because all I have to do is to compile before the validation therefore if there are a references that cannot be resolved yet (eg. cross-refering types), the last compile statement right before `validator(value)` should throw.